Modules/_sqlite/connection.c

```diff
@@ -449,7 +451,9 @@ connection_close(pysqlite_Connection *self)
             if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
                 remove_callbacks(self->db);
             }
-            (void)connection_exec_stmt(self, "ROLLBACK");
+            if (connection_exec_stmt(self, "ROLLBACK") < 0) {
+                PyErr_WriteUnraisable((PyObject *)self);
```

Comment:
Does not it contradict the comment above?

Reply:
Not sure that it is right to swallow all errors here. `Connection.close()` and `Connection.__init__()` can propagate it to the caller. Only destructor cannot do this, but `PyErr_WriteUnraisable()` is already called for exceptions raised in destructors.

Reply:
> Does not it contradict the comment above?

If `PyErr_WriteUnraisable` does not handle being called during interpreter shutdown, we should `PyErr_Clear` instead.

> Not sure that it is right to swallow all errors here.

The aim of this PR is to add error handling where it is not present.

Reply:
Serhiy is correct, the close() method can report exceptions. It should not be logged as an "unraisable exception". I suggest reporting if an exception was raised to the caller, and the caller decides to logs an unraisable exception, or to report it to its own caller (return NULL aka "oh no, an exception was raised").

Reply:
> Serhiy is correct, the close() method can report exceptions. It should not be logged as an "unraisable exception". I suggest reporting if an exception was raised to the caller, and the caller decides to logs an unraisable exception, or to report it to its own caller (return NULL aka "oh no, an exception was raised").

Yeah, changing `connection_close` to return -1 if an exception was set sounds good to me. That's a nice improvement.

Reply:
I changed `connection_close` to return -1 if an exception is set, and the callers now decide if to propagate the exception or log an unraisable exception.

Regarding the other comment of Serhiy:

> Does not it contradict the comment above?

The comment in question is the one about not calling back into Python code if we're in the middle of interpreter tear-down. This can easily happen, for example just by registering an SQL statement trace callback and tearing down the interpreter without explicitly calling `close()` on the connection object. But ISTM we can call into Python code from the unraisable hook as well, so perhaps we are better off just clearing the exception in case of interpreter shutdown.